### DISCUSSION TITLE
Extension System

### DISCUSSION THREAD

**Discussion by sr229** [vignetteapp/vignette#216] [Ideas]
This has been requested by the community; however, this is kinda low priority as we focus most on the core components. The way this works is the following:

- Extensions can expose their settings in `MainMenu`.
- They will be strictly be conformant to the o!f model to properly load. This is considered "bare minimum" for what people requires to make an extension.
- They will be packaged as either a `.dll` or a `.nupkg` which the program can "extract" or "compile" into a DLL, something we can do once we have a better idea with how to dynamically load assemblies.

Anyone can propose a better design here since this is a RFC, we appreciate alternative approaches for this.

**Comment by LeNitrous** [comment]
Dynamically loading assemblies is covered by reflection via [`Assembly.Load`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load?view=net-5.0). We'll have to provide proper APIs for plugins to be able to be loaded beforehand.

**Comment by sr229** [comment]
> Dynamically loading assemblies is covered by reflection via [`Assembly.Load`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load?view=net-5.0). We'll have to provide proper APIs for plugins to be able to be loaded beforehand.

This is something we can provide using a NuGet package. We'll need to strive to make `vignette.Game` self-contained in a degree ala-`osu.Game`.
